src/drivers/mct/shr/seq_comm_mct.F90

```diff
@@ -917,27 +474,113 @@ subroutine seq_comm_init(Comm_in, nmlfile)
-
-    select case(esmf_logging)
-    case ("ESMF_LOGKIND_SINGLE")
-       esmf_logfile_kind = ESMF_LOGKIND_SINGLE
-    case ("ESMF_LOGKIND_MULTI")
-       esmf_logfile_kind = ESMF_LOGKIND_MULTI
-    case ("ESMF_LOGKIND_NONE")
-       esmf_logfile_kind = ESMF_LOGKIND_NONE
-    case default
-       if (mype == 0) then
-          write(logunit,*) trim(subname),' ERROR: Invalid value for esmf_logging, ',esmf_logging
-       endif
-       call shr_sys_abort(trim(subname)//' ERROR: Invalid value for esmf_logging '//esmf_logging)
-    end select
 
     call seq_comm_printcomms()
 
   end subroutine seq_comm_init
 
+  subroutine comp_comm_init(global_comm, comp_rootpe, comp_nthreads, comp_layout, &
+       comp_ntasks, comp_pestride, num_inst_comp, &
+       CPLID, COMPID, CPLCOMPID, ALLCOMPID, CPLALLCOMPID, name, count)
+    integer, intent(in) :: global_comm
+    integer, intent(in) :: comp_rootpe
+    integer, intent(in) :: comp_nthreads
+    character(len=*), intent(in) :: comp_layout
+    integer, intent(in) :: comp_ntasks
+    integer, intent(in) :: comp_pestride
+    integer, intent(in) :: num_inst_comp
+    integer, intent(in) :: CPLID
+    integer, intent(inout) :: COMPID(num_inst_comp)
+    integer, intent(inout) :: CPLCOMPID(num_inst_comp)
+    integer, intent(inout) :: ALLCOMPID
+    integer, intent(inout) :: CPLALLCOMPID
+    integer, intent(inout) :: count
+    character(len=*), intent(in) :: name
+
+    character(len=*), parameter :: subname = "comp_comm_init"
+    integer :: comp_inst_tasks
+    integer :: droot
+    integer :: current_task_rootpe
+    integer :: cmin(num_inst_comp), cmax(num_inst_comp), cstr(num_inst_comp)
+    integer :: n
+    integer :: pelist (3,1)
+    integer :: ierr
+    integer :: mype
+
+
+    call mpi_comm_rank(global_comm, mype, ierr)
+
+    count = count + 1
+    ALLCOMPID = count
+    count = count + 1
+    CPLALLCOMPID = count
+    do n = 1, num_inst_comp
+       count = count + 1
+       COMPID(n) = count
+       count = count + 1
+       CPLCOMPID(n) = count
+    enddo
+
+    if (mype == 0) then
+       !--- validation of inputs ---
+       ! rootpes >= 0
+       !! Determine the process layout
+       !!
+       !! We will assign comp_ntasks / num_inst_comp tasks to each component
+       !! instance.  (This may lead to unallocated tasks if comp_ntasks is
+       !! not an integer multiple of num_inst_comp.)
+
+       if (comp_rootpe < 0) then
+          call shr_sys_abort(trim(subname)//' ERROR: rootpes must be >= 0 for component '//name)
+       endif
+
+       if (trim(comp_layout) == trim(layout_concurrent)) then
+          comp_inst_tasks = comp_ntasks / num_inst_comp
+          droot = (comp_inst_tasks * comp_pestride)
+       elseif (trim(comp_layout) == trim(layout_sequential)) then
+          comp_inst_tasks = comp_ntasks
+          droot = 0
+       else
+          call shr_sys_abort(subname//' ERROR invalid comp_layout for component'//name)
```

Comment:
This is going to look funny without a space after component. Also, name should be trim(name) even though it is always being called with constants at the moment.